File: ADVANCED RECURSION/Tiling problem.cpp
#include<iostream>
using namespace std;

int tilingWays(int n){
    if(n==0){ // base case
        return 0;
    }
    if(n==1){ // base case
        return 1;
    }

    return tilingWays(n-1) + tilingWays(n-2); // n-1 for vertically tiles placed and n-2 for horizontally tiles placed

}

int main(){

    cout<< tilingWays(4) <<endl;
    return 0;
}

File: JAVA/Tiling problem.java
import java.util.*;

public class Main {
    public static int tiles(int n, int m){
        if(n == m){ // base case 1
            return 2;
        }
        
        if(n < m){ // base case 2
            return 1;
        }
        
        // vertically placed
        int vertPlaced = tiles(n-m, m);
        
        // horizontally placed
        int horzPlaced = tiles(n-1, m);
        
        return vertPlaced + horzPlaced;
    }
    
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int m = sc.nextInt();
		
		System.out.println(tiles(n, m));
	}
}
